src/headless_browser.ts

```diff
@@ -58,82 +67,215 @@ export type ExceptionDetails = { // exists when an error
 
   /**
-   * @param urlToVisit - The url to visit or open up
+   * To keep hold of our promises waiting for messages from the websocket
    */
-  constructor(urlToVisit: string) {
+  private resolvables: { [key: number]: Deferred<unknown> } = {};
+
+  constructor() {
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - PUBLIC ////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Build the headless browser
+   */
+  public async build() {
+    const paths = {
+      // deno-lint-ignore camelcase
+      windows_chrome_exe:
+        "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
+      // deno-lint-ignore camelcase
+      windows_chrome_exe_x86:
+        "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe",
+      darwin: "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
+      linux: "/usr/bin/google-chrome",
+    };
     let chromePath = "";
     switch (Deno.build.os) {
       case "darwin":
-        chromePath =
-          "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome";
+        chromePath = paths.darwin;
         break;
       case "windows":
-        chromePath = "start chrome";
-        break;
+        if (existsSync(paths.windows_chrome_exe)) {
+          chromePath = paths.windows_chrome_exe;
+          break;
+        }
+        if (existsSync(paths.windows_chrome_exe_x86)) {
+          chromePath = paths.windows_chrome_exe_x86;
+          break;
+        }
+        throw new Error(
+          "Cannot find path for chrome in windows. Submit an issue if you encounter this error",
+        );
       case "linux":
-        chromePath = "TODO"; // TODO(any) Finnd what the path for chrome is for linux
+        chromePath = paths.linux;
         break;
     }
     this.browser_process = Deno.run({
       cmd: [
         chromePath,
         "--headless",
-        "--virtual-time-budget=10000",
-        "--repl",
-        urlToVisit,
+        "--remote-debugging-port=9292",
+        "--disable-gpu",
+        "https://chromestatus.com",
       ],
-      stdin: "piped",
-      stdout: "piped",
-      stderr: "piped",
+      stderr: "piped", // so stuff isn't displayed in the terminal for the user
     });
+    // Wait until browser is ready
+    for await (
+      const line of readLines(this.browser_process.stderr as Deno.Reader)
+    ) {
+      if (line.indexOf("DevTools listening on ws://") > -1) {
+        break;
+      }
+    }
+    let debugUrl = "";
+    while (true) {
+      try {
+        const res = await fetch("http://localhost:9292/json/list");
+        const json = await res.json();
+        debugUrl = json[0]["webSocketDebuggerUrl"];
+        break;
+      } catch (err) {
+        // do nothing, loop again until the endpoint is ready
+      }
+    }
+    // Connect websocket
+    this.socket = new WebSocket(debugUrl);
```

Comment:
can we integrate `wocket`?

Reply:
Wocket is a ws server right? Sinco just needs a ws client - or do you mean to use-wocket-client?

Reply:
https://github.com/drashland/sinco/pull/19#issuecomment-739576066

same question i have for both of these comments... what's the best place i can read more about this?

Reply:
ill try get some info for you